src/chrome/internal/breakpoints/features/bpAtNotLoadedScriptViaHeuristicSetter.ts

```diff
@@ -0,0 +1,130 @@
+
+        const heuristicBPRecipe = mapToUrlRegexp(requestedBP, location.url.textRepresentation, location.position, runtimeLocation);
+
+        this._bprToHeuristicBPR.set(requestedBP, heuristicBPRecipe);
+        await this._targetBreakpoints.setBreakpointByUrlRegexp(heuristicBPRecipe, eventsConsumer);
+    }
+
+    public async removeBPRecipeIfNeeded(requestedBP: BPRecipeInSource): Promise<void> {
+        const heuristicBPR = this._bprToHeuristicBPR.tryGetting(requestedBP);
+        if (heuristicBPR !== undefined) {
+            await this._targetBreakpoints.removeBreakpoint(heuristicBPR);
+            this._bprToHeuristicBPR.delete(requestedBP);
+        }
+    }
+
+    private async getBPRInUrlRegexpPosition(requestedBP: BPRecipeInSource<IBPActionWhenHit>): Promise<LocationInUrl> {
+        const sourceIdentifier = requestedBP.location.resource.sourceIdentifier;
+        const sourceMap = await this._sourceMapTransformer.getSourceMapFromAuthoredPath(sourceIdentifier);
+        if (sourceMap !== null) {
+            const script = new SourceWithSourceMap(sourceMap);
+            const sourceMapper = new MappedSourcesMapper(script, sourceMap);
+            const mappedLocation = sourceMapper.getPositionInScript(requestedBP.location);
+            return new LocationInUrl(<IResourceIdentifier<CDTPScriptUrl>>sourceMap.generatedPath, mappedLocation.enclosingRange.range.start);
+        }
+
+        // We don't know if this cast is correct or not. If it's not, the breakpoint will not bind, as designed
+        return new LocationInUrl(<IResourceIdentifier<CDTPScriptUrl>>sourceIdentifier, requestedBP.location.position);
+    }
+}
+
+/**
+ * We use this file when we have a typescript file, for which we found a source-map using the EagerSourceMapTransformer before the correspondant script
+ * was loaded by CDTP, so we need to use the source-mapping logic, yet we don't have an instance of IScript to do it. We use this instance instead,
+ * that will correctly help map typescript files to javascript files for non .html files.
+ */
+export class SourceWithSourceMap implements IHasSourceMappingInformation {
+    public constructor(private readonly _sourceMap: SourceMap) { }
+
+    public get mappedSources(): IdentifiedLoadedSource<string>[] {
+        // It doesn't seem this method is needed when calling sourceMapper.getPositionInScript, so we don't need to implement it
+        throw new Error(`Not yet implemented: SourceWithSourceMap.mappedSources`);
+    }
+
+    public get startPositionInSource(): Position {
+        return Position.origin; // TODO: Try to figure out an heuristic for .html files
+    }
+
+    public toString(): string {
+        return `Source with map: ${this._sourceMap.generatedPath}`;
+    }
+}
+
+/**
+ * We need a runtime location to create a BPRecipeInRegexp. We don't have one, so we pass a this instance instead.
+ */
+class NoLoadedSourceAvailable implements ILoadedSource<CDTPScriptUrl> {
```

Comment:
Do we really need to create a dummy instance of this interface just to satisfy an arbitrary contract elsewhere in the code?  Would it be simpler to create an instance of `Location` that didn't care about whether we have a loaded source or not?

Reply:
Currently the type of the parameter we need to send is LocationInLoadedSource, so we'd need to pass something that satisfies that type...